### PR TITLE
Embed the settings inside the code

### PR DESCRIPTION
**Pull request by kceiw**
## Description

We used to have the settings in the json file and release it together in the module. We'll load and deserialize it when the module is imported. The change is to improve this process:
1. The json file isn't released in the module anymore.
2. The settings are embedded inside the code so we don't need to load and deserialize it at runtime.

The way to achieve it is to use the source generator in .NET 6. It reads the same json file and parse to get the field and value. Then generate the source that:
1. The property for the field/value in the json. This is the default settings.
2. The method to allow overriding the properties. This is used when we load the settings from the user's profile. This allows to override the default setting.

## Checklist

- [x] The title of the PR is clear and informative
- [x] I have read the [_Submitting Changes_](../blob/master/CONTRIBUTING.md#submitting-changes) section of [`CONTRIBUTING.md`](../blob/master/CONTRIBUTING.md)
- [ ] The appropriate `ChangeLog.md` file(s) has been updated:
    - For any service, the `ChangeLog.md` file can be found at `src/{{SERVICE}}/{{SERVICE}}/ChangeLog.md`
    - A snippet outlining the change(s) made in the PR should be written under the `## Upcoming Release` header -- no new version header should be added
- [x] The PR does not introduce [breaking changes](../blob/master/documentation/breaking-changes/breaking-changes-definition.md)
- [ ] If applicable, the changes made in the PR have proper test coverage
- [ ] For public API changes to cmdlets:
    - [ ] a cmdlet design review was approved for the changes in [this repository](https://github.com/Azure/azure-powershell-cmdlet-review-pr) (_Microsoft internal only_)
        - {Please put the link here}
    - [ ] the markdown help files have been regenerated using the commands listed [here](../blob/master/documentation/development-docs/help-generation.md#updating-all-markdown-files-in-a-module)

### CODE REVIEW

File: tools/Az.Tools.Predictor/Az.Tools.Predictor.SourceGenerator/SettingsSourceGenerator.cs [hunk 1]
```diff
@@ -0,0 +1,185 @@
+﻿using Microsoft.CodeAnalysis;
```
Comment (mirdaki) [inline tools/Az.Tools.Predictor/Az.Tools.Predictor.SourceGenerator/SettingsSourceGenerator.cs:1]: Nit: Add MS copyright?

**Review by mirdaki — DISMISSED** [review]
Looks good. Just a nit

**Review by mirdaki — APPROVED** [review]

File: tools/Az.Tools.Predictor/Az.Tools.Predictor.SourceGenerator/SettingsSourceGenerator.cs [hunk 2]
```diff
@@ -0,0 +1,199 @@
...
+///
+/// It gathers the key:value pair from the json file, and generates the public properties accordingly.
+/// The key from the json is the property name and the value from the json is the property value. It infers
+/// the property type based on the value.
+/// The generated properties are the default values of the settings.
+/// In addition to generating the properties, it also generate the method to get the settings from the user's profile,
+/// and override the property accordingly if it's set in the user's profile setting.
+/// </summary>
+[Generator]
+public class SettingsSourceGenerator : ISourceGenerator
```
Comment (jjaguirre394) [inline tools/Az.Tools.Predictor/Az.Tools.Predictor.SourceGenerator/SettingsSourceGenerator.cs:46]: I just learned about this feature of dotnet. Thanks for doing this, really cool! :)

File: tools/Az.Tools.Predictor/Az.Tools.Predictor.SourceGenerator/SettingsSourceGenerator.cs [hunk 3]
```diff
@@ -0,0 +1,199 @@
...
+        var fileContent = File.ReadAllText(settingFilePath, Encoding.UTF8);
+        var jsonDocument = JsonDocument.Parse(fileContent);
+        var jsonRoot = jsonDocument.RootElement;
+        var settingProperties = new List<SettingProperty>();
+
+        // Parse the json file and gather the property information, including name, type, and value.
+
+        foreach (var element in jsonRoot.EnumerateObject())
+        {
+            var addedProperty = new SettingProperty()
```
Comment (jjaguirre394) [inline tools/Az.Tools.Predictor/Az.Tools.Predictor.SourceGenerator/SettingsSourceGenerator.cs:96]: nit: maybe something like "newProperty" or "generatedProperty"? I'm okay if it stays as addedProperty, just found myself reading this as new or generated property.

File: tools/Az.Tools.Predictor/Az.Tools.Predictor.SourceGenerator/SettingsSourceGenerator.cs [hunk 4]
```diff
@@ -0,0 +1,199 @@
...
+            {
+                Name = element.Name.Substring(0, 1).ToUpperInvariant() + element.Name.Substring(1),
+            };
+
+            // There are only three types used in the json file: bool, int, and string.
```
Comment (jjaguirre394) [inline tools/Az.Tools.Predictor/Az.Tools.Predictor.SourceGenerator/SettingsSourceGenerator.cs:101]: We support doubles as well right?

File: tools/Az.Tools.Predictor/Az.Tools.Predictor.SourceGenerator/Az.Tools.Predictor.SourceGenerator.csproj [hunk 5]
```diff
@@ -0,0 +1,47 @@
...
+  </ItemGroup>
+
+  <PropertyGroup>
+    <GetTargetPathDependsOn>$(GetTargetPathDependsOn);GetDependencyTargetPaths</GetTargetPathDependsOn>
+  </PropertyGroup>
+
+  <Target Name="GetDependencyTargetPaths" AfterTargets="ResolvePackageDependenciesForBuild">
+    <ItemGroup>
+      <!-- See potential transient dependency workaround here https://turnerj.com/blog/the-pain-points-of-csharp-source-generators -->
+      <TargetPathWithTargetPlatformMoniker Include="@(ResolvedCompileFileDefinitions)" IncludeRuntimeDependency="false" />
```
Comment (jjaguirre394) [inline tools/Az.Tools.Predictor/Az.Tools.Predictor.SourceGenerator/Az.Tools.Predictor.SourceGenerator.csproj:44]: Maybe a small comment here describing the issue that this workaround is solving as I am sure we will run into it again if we use this generation feature haha

**Review by jjaguirre394 — DISMISSED** [review]
Really awesome dotnet code generation feature. Thanks for doing this!

File: tools/Az.Tools.Predictor/Az.Tools.Predictor.SourceGenerator/SettingsSourceGenerator.cs [hunk 6]
```diff
@@ -0,0 +1,199 @@
...
+// ----------------------------------------------------------------------------------
+
+namespace {settingClass.ContainingNamespace}
+{{
+    // This file is generated from the source generator. DO NOT edit this file.
+    sealed partial class {settingClass.Name}
+    {{
+");
+
+        var settingFilePath = context.AdditionalFiles.Single((file) => file.Path.EndsWith("AzPredictorSettings.json", StringComparison.Ordinal)).Path;
```
Comment (baopingz) [inline tools/Az.Tools.Predictor/Az.Tools.Predictor.SourceGenerator/SettingsSourceGenerator.cs:85]: nit: settingsFilePath

**Review by baopingz — APPROVED** [review]

**Review by mirdaki — APPROVED** [review]

**Review by dingmeng-xue — APPROVED** [review]

**Pull request merged**